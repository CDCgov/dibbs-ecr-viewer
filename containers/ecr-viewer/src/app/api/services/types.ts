import {
  ColumnType,
  Generated,
  Insertable,
  Selectable,
  Updateable,
} from "kysely";

export interface ecr_data {
  eICR_ID: Generated<string>;
  set_id: string | undefined;
  data_source: string | undefined;
  fhir_reference_link: string | undefined;
  eicr_version_number: string | undefined;

  patient_name_first: string;
  patient_name_last: string;
  patient_birth_date: ColumnType<Date, string | undefined, never>;

  date_created: ColumnType<Date, string | undefined, never>;
  report_date: ColumnType<Date, string>;
}

export interface ecr_rr_conditions {
  uuid: Generated<string>;
  eICR_ID: string;
  condition: string;
}

export interface ecr_rr_rule_summaries {
  uuid: Generated<string>;
  ecr_rr_conditions_id: string;
  rule_summary: string;
}

export type ECR = Selectable<ecr_data>;
export type NewECR = Insertable<ecr_data>;
export type ECRUpdate = Updateable<ecr_data>;

export type ECRConditions = Selectable<ecr_rr_conditions>;
export type NewECRConditions = Insertable<ecr_rr_conditions>;
export type ECRConditionsUpdate = Updateable<ecr_rr_conditions>;

export type ECRRuleSummaries = Selectable<ecr_rr_rule_summaries>;
export type NewECRRuleSummaries = Insertable<ecr_rr_rule_summaries>;
export type ECRRuleSummariesUpdate = Updateable<ecr_rr_rule_summaries>;

export interface Core {
  ecr_data: ecr_data;
  ecr_rr_conditions: ecr_rr_conditions;
  ecr_rr_rule_summaries: ecr_rr_rule_summaries;
}
