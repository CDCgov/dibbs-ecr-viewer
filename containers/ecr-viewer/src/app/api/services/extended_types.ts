import {
  ColumnType,
  Generated,
  Insertable,
  Selectable,
  Updateable,
} from "kysely";

export interface ecr_data {
  eICR_ID: Generated<string>;
  set_id: string | undefined;
  fhir_reference_link: string | null;
  last_name: string | undefined;
  first_name: string | undefined;
  birth_date: ColumnType<Date | undefined>;
  gender: string | undefined;
  birth_sex: string | undefined;
  gender_identity: string | undefined;
  race: string | undefined;
  ethnicity: string | undefined;
  latitude: number | undefined;
  longitude: number | undefined;
  homelessness_status: string | undefined;
  disabilities: string | undefined;
  tribal_affiliation: string | undefined;
  tribal_enrollment_status: string | undefined;
  current_job_title: string | undefined;
  current_job_industry: string | undefined;
  usual_occupation: string | undefined;
  usual_industry: string | undefined;
  preferred_language: string | undefined;
  pregnancy_status: string | undefined;
  rr_id: string | undefined;
  processing_status: string | undefined;
  eicr_version_number: string | undefined;
  authoring_date: ColumnType<Date | undefined>;
  authoring_provider: string | undefined;
  provider_id: string | undefined;
  facility_id: string | undefined;
  facility_name: string | undefined;
  encounter_type: string | undefined;
  encounter_start_date: ColumnType<Date | undefined>;
  encounter_end_date: ColumnType<Date | undefined>;
  reason_for_visit: string | undefined;
  active_problems: string[] | undefined;
  date_created: ColumnType<Date | undefined>;
}

export interface patient_address {
  uuid: Generated<string>;
  use: "home" | "work" | "temp" | "old" | "billing" | undefined;
  type: "postal" | "physical" | "both" | undefined;
  text: string | undefined;
  line: string[] | undefined;
  city: string | undefined;
  district: string | undefined;
  state: string | undefined;
  postal_code: string | undefined;
  country: string | undefined;
  period_start: ColumnType<Date | undefined>;
  period_end: ColumnType<Date | undefined>;
  eICR_ID: string | null; // Nullable foreign key reference
}

export interface ecr_labs {
  uuid: Generated<string>;
  eICR_ID: string;
  test_type: string | undefined;
  test_type_code: string | undefined;
  test_type_system: string | undefined;
  test_result_qualitative: string | undefined;
  test_result_quantitative: number | null;
  test_result_units: string | undefined;
  test_result_code: string | undefined;
  test_result_code_display: string | undefined;
  test_result_code_system: string | undefined;
  test_result_interpretation: string | undefined;
  test_result_interpretation_code: string | undefined;
  test_result_interpretation_system: string | undefined;
  test_result_reference_range_low_value: number | string | null;
  test_result_reference_range_low_units: string | undefined;
  test_result_reference_range_high_value: number | string | null;
  test_result_reference_range_high_units: string | undefined;
  specimen_type: string | undefined;
  specimen_collection_date: ColumnType<Date | undefined>;
  performing_lab: string | undefined;
}

export interface ecr_rr_conditions {
  uuid: Generated<string>;
  eICR_ID: string;
  condition: string;
}

export interface ecr_rr_rule_summaries {
  uuid: Generated<string>;
  ecr_rr_conditions_id: string;
  rule_summary: string;
}

export type ExtendedECR = Selectable<ecr_data>;
export type NewExtendedECR = Insertable<ecr_data>;
export type ExtendedECRUpdate = Updateable<ecr_data>;

export type PatientAddress = Selectable<patient_address>;
export type NewPatientAddress = Insertable<patient_address>;
export type PatientAddressUpdate = Updateable<patient_address>;

export type ECRLabs = Selectable<ecr_labs>;
export type NewECRLabs = Insertable<ecr_labs>;
export type ECRLabsUpdate = Updateable<ecr_labs>;

export type ECRConditions = Selectable<ecr_rr_conditions>;
export type NewECRConditions = Insertable<ecr_rr_conditions>;
export type ECRConditionsUpdate = Updateable<ecr_rr_conditions>;

export type ECRRuleSummaries = Selectable<ecr_rr_rule_summaries>;
export type NewECRRuleSummaries = Insertable<ecr_rr_rule_summaries>;
export type ECRRuleSummariesUpdate = Updateable<ecr_rr_rule_summaries>;

export interface Extended {
  ecr_data: ecr_data;
  patient_address: patient_address;
  ecr_labs: ecr_labs;
  ecr_rr_conditions: ecr_rr_conditions;
  ecr_rr_rule_summaries: ecr_rr_rule_summaries;
}
