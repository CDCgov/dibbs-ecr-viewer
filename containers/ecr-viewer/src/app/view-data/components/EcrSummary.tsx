import React from "react";
import {
  DataDisplay,
  DataTableDisplay,
  DisplayDataProps,
} from "@/app/view-data/components/DataDisplay";
import { Accordion, Tag } from "@trussworks/react-uswds";
import { AccordionItem } from "@/app/view-data/types";
import { toKebabCase } from "@/app/utils/format-utils";

interface EcrSummaryProps {
  patientDetails: DisplayDataProps[];
  encounterDetails: DisplayDataProps[];
  conditionSummary: ConditionSummary[];
  snomed?: string;
}

export interface ConditionSummary {
  title: string;
  snomed: string;
  conditionDetails: DisplayDataProps[];
  clinicalDetails: DisplayDataProps[];
  labDetails: DisplayDataProps[];
  immunizationDetails: DisplayDataProps[];
}

/**
 * Generates a JSX element to display eCR viewer summary
 * @param props - Properties for the eCR Viewer Summary section
 * @param props.patientDetails - Array of title and values to be displayed in patient details section
 * @param props.encounterDetails - Array of title and values to be displayed in encounter details section
 * @param props.conditionSummary - Array of condition details
 * @param props.snomed - SNOMED code being requested
 * @returns a react element for ECR Summary
 */
const EcrSummary: React.FC<EcrSummaryProps> = ({
  patientDetails,
  encounterDetails,
  conditionSummary,
  snomed,
}) => {
  const conditionSummaryAccordionItems: AccordionItem[] = conditionSummary.map(
    (condition) => {
      return {
        title: condition.title,
        id: toKebabCase(condition.title),
        headingLevel: "h4",
        className: "side-nav-ignore border-1px border-accent-cool-darker",
        expanded: snomed === condition.snomed || conditionSummary.length === 1,
        content: (
          <>
            {condition.conditionDetails.map((item) => (
              <DataDisplay item={item} key={item.title} />
            ))}
            <h5
              className="text-bold margin-top-0 margin-bottom-1"
              id="relevant-clinical"
            >
              Clinical Sections Relevant to Reportable Condition
            </h5>
            {condition.immunizationDetails.length > 0 && (
              <div className={"margin-top-0"}>
                {condition.immunizationDetails.map((item, index) => (
                  <DataTableDisplay item={item} key={index} />
                ))}
              </div>
            )}
            <div className={"margin-top-0"}>
              {condition.clinicalDetails.map((item, index) => (
                <DataTableDisplay item={item} key={`${item.title}-${index}`} />
              ))}
            </div>
            <h5
              className="text-bold margin-0 margin-bottom-1"
              id="relevant-labs"
            >
              Lab Results Relevant to Reportable Condition
            </h5>
            <div className="margin-top-0">
              {condition.labDetails.map((item, index) => (
                <DataTableDisplay item={item} key={`${item.title}-${index}`} />
              ))}
            </div>
          </>
        ),
      };
    },
  );
  return (
    <div
      className="usa-summary-box padding-3"
      aria-labelledby="summary-box-key-information"
    >
      <div className="usa-summary-box__body margin-bottom-05">
        <h3
          className="summary-box-key-information side-nav-ignore"
          id={"patient-summary"}
        >
          Patient Summary
        </h3>
        <div className="usa-summary-box__text">
          {patientDetails.map((item) => (
            <DataDisplay item={item} key={item.title} themeColor="blue" />
          ))}
        </div>
      </div>
      <div className="usa-summary-box__body">
        <h3
          className="summary-box-key-information side-nav-ignore"
          id="encounter-summary"
        >
          Encounter Summary
        </h3>
        <div className="usa-summary-box__text">
          {encounterDetails.map((item) => (
            <DataDisplay item={item} key={item.title} themeColor="blue" />
          ))}
        </div>
      </div>
      <div className="usa-summary-box__body">
        <h3
          className={
            "summary-box-key-information side-nav-ignore header-with-tag"
          }
          id={"condition-summary"}
        >
          <div>Condition Summary</div>
          <div>
            <Tag className="tag-info">
              {numConditionsText(conditionSummaryAccordionItems.length)}
            </Tag>
          </div>
        </h3>
        <div className="usa-summary-box__text condition-details-accordion">
          <Accordion items={conditionSummaryAccordionItems} />
        </div>
      </div>
    </div>
  );
};

/**
 * Returns a formatted string indicating the number of reportable conditions.
 * @param numConditions - Number of conditions.
 * @returns A formatted string that specifies the number of conditions found.
 */
const numConditionsText = (numConditions: number) => {
  return numConditions === 1
    ? `${numConditions} CONDITION FOUND`
    : `${numConditions} CONDITIONS FOUND`;
};

export default EcrSummary;
