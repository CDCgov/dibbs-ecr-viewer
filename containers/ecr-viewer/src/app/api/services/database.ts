// Kysely ORM Connection Client

import { Kysely, PostgresDialect, MssqlDialect } from "kysely";
import { Pool } from "pg";
import * as tarn from "tarn";
import * as tedious from "tedious";

import { Extended } from "./extended_types";
import { Core } from "./types";

// Dialect to communicate with the database, interface to define its structure.

let db: Kysely<Core> | Kysely<Extended>;

if (process.env.METADATA_DATABASE_TYPE === "sqlserver") {
  db = new Kysely<Extended>({
    dialect: new MssqlDialect({
      tarn: {
        ...tarn,
        options: {
          min: 0,
          max: 100,
        },
      },
      tedious: {
        ...tedious,
        connectionFactory: () =>
          new tedious.Connection({
            authentication: {
              options: {
                password: process.env.SQL_SERVER_PASSWORD,
                userName: process.env.SQL_SERVER_USER || "sa",
              },
              type: "default",
            },
            options: {
              database: "master",
              port: 1433,
              trustServerCertificate: true,
              connectTimeout: 30000,
            },
            server: process.env.SQL_SERVER_HOST || "localhost",
          }),
      },
    }),
  });
} else {
  db = new Kysely<Core>({
    dialect: new PostgresDialect({
      pool: new Pool({
        database: process.env.POSTGRES_DATABASE || "ecr_viewer_db",
        host: process.env.POSTGRES_HOST,
        user: process.env.POSTGRES_USER,
        port: parseInt(process.env.POSTGRES_PORT || "5432"),
        max: parseInt(process.env.POSTGRES_MAX_THREADPOOL || "10"),
      }),
    }),
  });
}

export { db };
