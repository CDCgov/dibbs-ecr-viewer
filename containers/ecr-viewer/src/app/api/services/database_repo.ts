import { Kysely } from "kysely";

import { db } from "./database";
import {
  ECRConditions,
  NewECRConditions,
  ECRConditionsUpdate,
  ECRRuleSummaries,
  NewECRRuleSummaries,
  ECRRuleSummariesUpdate,
  ECR,
  NewECR,
  ECRUpdate,
} from "./types";
import { Core } from "./types";

// ECR_DATA

/**
 * Finds an eICR by its ID
 * @async
 * @function findEcrById
 * @param id - thd ID of the eCR to be looked up
 * @returns an eICR object
 */
export async function findEcrById(id: string | null) {
  if (!id) {
    return console.error("eICR ID is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .selectFrom("ecr_data")
      .where("eICR_ID", "=", id)
      .selectAll()
      .executeTakeFirst();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Finds an eICR by its criteria
 * @async
 * @function findEcr
 * @param criteria - the Partial<ECR> used to filter against
 * @returns an eICR object
 */
export async function findEcr(criteria: Partial<ECR> | null) {
  let query = (db as Kysely<Core>).selectFrom("ecr_data");

  if (!criteria || criteria === null) {
    return console.error("eICR Criteria is required.");
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.set_id) {
    query = query.where("set_id", "=", criteria.set_id);
  }

  if (criteria.data_source) {
    query = query.where("data_source", "=", criteria.data_source);
  }

  if (criteria.fhir_reference_link) {
    query = query.where(
      "fhir_reference_link",
      "=",
      criteria.fhir_reference_link,
    );
  }

  if (criteria.patient_name_last !== undefined) {
    query = query.where(
      "patient_name_last",
      criteria.patient_name_last === null ? "is" : "=",
      criteria.patient_name_last,
    );
  }

  if (criteria.patient_birth_date) {
    query = query.where("patient_birth_date", "=", criteria.patient_birth_date);
  }

  if (criteria.date_created) {
    query = query.where("date_created", "=", criteria.date_created);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eICR object
 * @async
 * @function createEcr
 * @param ecr - the NewECR to be persisted
 * @returns the created eICR object
 */
export async function createEcr(ecr: NewECR | null) {
  if (!ecr || ecr === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_data")
      .values(ecr)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eICR object
 * @async
 * @function updateEcr
 * @param eICR_ID - the eICR_ID of the eCR to be updated
 * @param updateWith - the ECRUpdate to be applied to the existing record
 * @returns the updated eICR object
 */
export async function updateEcr(eICR_ID: string | null, updateWith: ECRUpdate) {
  await (db as Kysely<Core>)
    .updateTable("ecr_data")
    .set(updateWith)
    .where("eICR_ID", "=", eICR_ID)
    .execute();
}

/**
 * Deletes an eICR object
 * @async
 * @function deleteEcr
 * @param eICR_ID - the eICR_ID of the record to be deleted
 * @returns the deleted eICR object
 */
export async function deleteEcr(eICR_ID: string | null) {
  const ecr = await findEcrById(eICR_ID);

  if (ecr) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_data")
      .where("eICR_ID", "=", eICR_ID)
      .execute();
  }

  return ecr;
}

// ECR_RR_CONDITIONS

/**
 * Finds an eCR condition by its ID
 * @async
 * @function findEcrConditionById
 * @param id - the UUID of the eCR condition being looked up
 * @returns an eCR condition object
 */
export async function findEcrConditionById(id: string) {
  return await (db as Kysely<Core>)
    .selectFrom("ecr_rr_conditions")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR condition by its criteria
 * @async
 * @function findEcrCondition
 * @param criteria - the Partial<ECRConditions> filtering criteria to be applied
 * @returns an eCR condition object
 */
export async function findEcrCondition(criteria: Partial<ECRConditions>) {
  let query = (db as Kysely<Core>).selectFrom("ecr_rr_conditions");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.condition) {
    query = query.where("condition", "=", criteria.condition);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR condition object
 * @async
 * @function createEcrCondition
 * @param condition - the NewECRConditions to be created
 * @returns the created eCR condition object
 */
export async function createEcrCondition(condition: NewECRConditions | null) {
  if (!condition || condition === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_rr_conditions")
      .values(condition)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR condition object
 * @async
 * @function updateEcrCondition
 * @param uuid - the UUID of the eCR Condition to be updated
 * @param updateWith - the ECRConditionsUpdate to be applied
 * @returns the updated eCR condition object
 */
export async function updateEcrCondition(
  uuid: string,
  updateWith: ECRConditionsUpdate,
) {
  await (db as Kysely<Core>)
    .updateTable("ecr_rr_conditions")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR condition object
 * @async
 * @function deleteEcrCondition
 * @param uuid - thge UUID of the record to be deleted
 * @returns the deleted eCR condition object
 */
export async function deleteEcrCondition(uuid: string) {
  const ecr = await findEcrConditionById(uuid);

  if (ecr) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_rr_conditions")
      .where("uuid", "=", uuid)
      .execute();
  }

  return ecr;
}

// ECR_RR_RULE_SUMMARIES

/**
 * Finds an eCR rule summary by its ID
 * @async
 * @function findEcrRuleById
 * @param id - the ID of the eCR being looked up
 * @returns an eCR rule object
 */
export async function findEcrRuleById(id: string) {
  return await (db as Kysely<Core>)
    .selectFrom("ecr_rr_rule_summaries")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR rule summary by its criteria
 * @async
 * @function findEcrRule
 * @param criteria - the Partial<ECRRuleSummaries> filtering criteria
 * @returns an eCR rule object
 */
export async function findEcrRule(criteria: Partial<ECRRuleSummaries>) {
  let query = (db as Kysely<Core>).selectFrom("ecr_rr_rule_summaries");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.ecr_rr_conditions_id) {
    query = query.where(
      "ecr_rr_conditions_id",
      "=",
      criteria.ecr_rr_conditions_id,
    );
  }

  if (criteria.rule_summary) {
    query = query.where("rule_summary", "=", criteria.rule_summary);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR rule summary object
 * @async
 * @param rule_summary - the NewECRRuleSummaries record to be created
 * @function createEcrRule
 * @returns the created eCR rule object
 */
export async function createEcrRule(rule_summary: NewECRRuleSummaries) {
  if (!rule_summary || rule_summary === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_rr_rule_summaries")
      .values(rule_summary)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR rule summary object
 * @async
 * @function updateEcrRule
 * @param uuid - the UUID of the eCR rule to be updated
 * @param updateWith - the ECRRuleSummariesUpdate to be applied to the record
 * @returns the updated eCR rule object
 */
export async function updateEcrRule(
  uuid: string,
  updateWith: ECRRuleSummariesUpdate,
) {
  await (db as Kysely<Core>)
    .updateTable("ecr_rr_rule_summaries")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR rule summary object
 * @async
 * @function deleteEcrRule
 * @param uuid - the UUID of the eCR rule to be deleted
 * @returns the deleted eCR rule object
 */
export async function deleteEcrRule(uuid: string) {
  const rule_summary = await findEcrRuleById(uuid);

  if (rule_summary) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_rr_rule_summaries")
      .where("uuid", "=", uuid)
      .execute();
  }

  return rule_summary;
}
