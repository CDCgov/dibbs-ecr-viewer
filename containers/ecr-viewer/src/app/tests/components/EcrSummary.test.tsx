import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";

import { findCurrentAddress } from "@/app/services/ecrSummaryService";
import EcrSummary, {
  ConditionSummary,
} from "@/app/view-data/components/EcrSummary";

describe("EcrSummary tests", () => {
  describe("EcrSummary", () => {
    const patientDetails = [
      {
        title: "Patient Name",
        value: "Han Solo",
      },
      {
        title: "DOB",
        value: "04/15/1900",
      },
      {
        title: "Sex",
        value: "male",
      },
      {
        title: "Patient Address",
        value: "1050 Millennium Falcon ST\nSpace, CA\n00123, US",
      },
      {
        title: "Patient Contact",
        value: "Home 555-555-55555\nHANSOLOFAKEEMAIL@EXAMPLE.COM",
      },
    ];
    const encounterDetails = [
      {
        title: "Facility Name",
        value: "Millennium Falcon Med Bay",
      },
      {
        title: "Facility Contact",
        value: "555-555-5555",
      },
      {
        title: "Encounter Date/Time",
        value: "Start: 05/13/2000 7:25 AM UTC\nEnd: 05/13/2000 9:57 AM UTC",
      },
      {
        title: "Encounter Type",
        value: "Emergency",
      },
    ];
    const covidConditionDetails: ConditionSummary[] = [
      {
        title: "Influenza caused by Influenza A virus subtype H5N1 (disorder)",
        snomed: "test-snomed-123",
        conditionDetails: [
          {
            title: "RCKMS Rule Summary",
            value: "covid summary",
          },
        ],
        clinicalDetails: [
          {
            title: "Relevant Clinical",
            value: "covid clinical",
          },
        ],
        labDetails: [
          {
            title: "Relevant Labs",
            value: "covid lab",
          },
        ],
        immunizationDetails: [
          {
            title: "Relevant immmunizations",
            value: "table would be here",
            dividerLine: true,
          },
        ],
      },
    ];
    const hepConditionDetails: ConditionSummary[] = [
      {
        title: "Hep C",
        snomed: "test-snomed-456",
        conditionDetails: [
          {
            title: "RCKMS Rule Summary",
            value: "hep c summary",
          },
        ],
        clinicalDetails: [
          {
            title: "Relevant Clinical",
            value: "hep c clinical",
          },
        ],
        labDetails: [
          {
            title: "Relevant Labs",
            value: "hep c lab",
          },
        ],
        immunizationDetails: [],
      },
    ];

    beforeAll(() => {});
    it("should match snapshot", () => {
      const { container } = render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={covidConditionDetails}
        />,
      );

      expect(container).toMatchSnapshot();
    });
    it("should pass accessibility test", async () => {
      const { container } = render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={covidConditionDetails}
        />,
      );

      expect(await axe(container)).toHaveNoViolations();
    });
    it("should open the condition details when there is one", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={covidConditionDetails}
        />,
      );

      expect(screen.getByText("covid summary")).toBeVisible();
    });
    it("should open the condition when the snomed matches", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={[...covidConditionDetails, ...hepConditionDetails]}
          snomed="test-snomed-456"
        />,
      );

      expect(screen.getByText("hep c summary")).toBeVisible();
    });
    it("should open no condition details when there is many and no match", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={[...covidConditionDetails, ...hepConditionDetails]}
        />,
      );

      expect(screen.getByText("hep c summary")).not.toBeVisible();
      expect(screen.getByText("covid summary")).not.toBeVisible();
    });
    it("should show 0 reportable conditions tag", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={[]}
        />,
      );

      expect(screen.getByText("0 CONDITIONS FOUND"));
    });
    it("should show 1 reportable condition tag", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={covidConditionDetails}
        />,
      );

      expect(screen.getByText("1 CONDITION FOUND"));
    });
    it("should show 2 reportable conditions tag", () => {
      render(
        <EcrSummary
          patientDetails={patientDetails}
          encounterDetails={encounterDetails}
          conditionSummary={[...covidConditionDetails, ...hepConditionDetails]}
        />,
      );

      expect(screen.getByText("2 CONDITIONS FOUND"));
    });
  });

  describe("findCurrentAddress", () => {
    const base = {
      line: ["123 Main St"],
    };

    it("should return empty when no addresses available", () => {
      const actual = findCurrentAddress([]);

      expect(actual).toEqual("");
    });

    it("should return first address when no use or period", () => {
      const actual = findCurrentAddress([
        { ...base, city: "1" },
        { ...base, city: "2" },
      ]);

      expect(actual).toEqual("123 Main St\n1");
    });

    it("should return first home address when no current period", () => {
      const actual = findCurrentAddress([
        { ...base, use: "work", city: "1" },
        { ...base, use: "home", city: "2" },
        { ...base, use: "home", city: "3" },
        {
          ...base,
          use: "home",
          city: "3",
          period: { start: "2020-03-01", end: "2020-04-01" },
        },
      ]);

      expect(actual).toEqual("123 Main St\n2");
    });

    it("should return current home address", () => {
      const actual = findCurrentAddress([
        { ...base, use: "work", city: "1" },
        { ...base, use: "home", city: "2" },
        { ...base, use: "home", city: "3", period: { start: "2024-03-13" } },
        {
          ...base,
          use: "home",
          city: "4",
          period: { start: "2024-03-10", end: "2024-03-12" },
        },
      ]);

      expect(actual).toEqual("123 Main St\n3");
    });
  });
});
