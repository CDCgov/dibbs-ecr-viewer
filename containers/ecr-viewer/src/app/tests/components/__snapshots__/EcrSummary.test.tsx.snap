// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcrSummary tests EcrSummary should match snapshot 1`] = `
<div>
  <div
    aria-labelledby="summary-box-key-information"
    class="usa-summary-box padding-3"
  >
    <div
      class="usa-summary-box__body margin-bottom-05"
    >
      <h3
        class="summary-box-key-information side-nav-ignore"
        id="patient-summary"
      >
        Patient Summary
      </h3>
      <div
        class="usa-summary-box__text"
      >
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Patient Name
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              Han Solo
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              DOB
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              04/15/1900
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Sex
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              male
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Patient Address
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              1050 Millennium Falcon ST
Space, CA
00123, US
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Patient Contact
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              Home 555-555-55555
HANSOLOFAKEEMAIL@EXAMPLE.COM
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
      </div>
    </div>
    <div
      class="usa-summary-box__body"
    >
      <h3
        class="summary-box-key-information side-nav-ignore"
        id="encounter-summary"
      >
        Encounter Summary
      </h3>
      <div
        class="usa-summary-box__text"
      >
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Facility Name
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              Millennium Falcon Med Bay
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Facility Contact
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              555-555-5555
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Encounter Date/Time
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              Start: 05/13/2000 7:25 AM UTC
End: 05/13/2000 9:57 AM UTC
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
        <div>
          <div
            class="grid-row"
          >
            <div
              class="data-title padding-right-1"
            >
              Encounter Type
            </div>
            <div
              class="grid-col maxw7 text-pre-line p-list"
            >
              Emergency
            </div>
          </div>
          <div
            class="section__line_blue"
          />
        </div>
      </div>
    </div>
    <div
      class="usa-summary-box__body"
    >
      <h3
        class="summary-box-key-information side-nav-ignore header-with-tag"
        id="condition-summary"
      >
        <div>
          Condition Summary
        </div>
        <div>
          <span
            class="usa-tag tag-info"
            data-testid="tag"
          >
            1 CONDITION FOUND
          </span>
        </div>
      </h3>
      <div
        class="usa-summary-box__text condition-details-accordion"
      >
        <div
          class="usa-accordion"
          data-testid="accordion"
        >
          <h4
            class="usa-accordion__heading side-nav-ignore border-1px border-accent-cool-darker"
          >
            <button
              aria-controls="influenza-caused-by-influenza-a-virus-subtype-h5n1-disorder"
              aria-expanded="true"
              class="usa-accordion__button"
              data-testid="accordionButton_influenza-caused-by-influenza-a-virus-subtype-h5n1-disorder"
              type="button"
            >
              Influenza caused by Influenza A virus subtype H5N1 (disorder)
            </button>
          </h4>
          <div
            class="usa-accordion__content usa-prose side-nav-ignore border-1px border-accent-cool-darker"
            data-testid="accordionItem_influenza-caused-by-influenza-a-virus-subtype-h5n1-disorder"
            id="influenza-caused-by-influenza-a-virus-subtype-h5n1-disorder"
          >
            <div>
              <div
                class="grid-row"
              >
                <div
                  class="data-title padding-right-1"
                >
                  RCKMS Rule Summary
                </div>
                <div
                  class="grid-col maxw7 text-pre-line p-list"
                >
                  covid summary
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
            <h5
              class="text-bold margin-top-0 margin-bottom-1"
              id="relevant-clinical"
            >
              Clinical Sections Relevant to Reportable Condition
            </h5>
            <div
              class="margin-top-0"
            >
              <div
                class="grid-row"
              >
                <div
                  class="grid-col-auto width-full text-pre-line"
                >
                  table would be here
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
            <div
              class="margin-top-0"
            >
              <div
                class="grid-row"
              >
                <div
                  class="grid-col-auto width-full text-pre-line"
                >
                  covid clinical
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
            <div
              class="text-bold margin-0 margin-bottom-1"
              id="relevant-labs"
            >
              Lab Results Relevant to Reportable Condition
            </div>
            <div
              class="margin-top-0"
            >
              <div
                class="grid-row"
              >
                <div
                  class="grid-col-auto width-full text-pre-line"
                >
                  covid lab
                </div>
              </div>
              <div
                class="section__line_gray"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
